### PR TITLE
chore: refacto reference generation

### PR DESCRIPTION
**Pull request by JacquesDurand**
| Q             | A
| ------------- | ---
| Branch?       | docs
| Bug fix?      | no 
| New feature?  | no
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | 

Refacto reference generation using symfony commands:

- [x] Create a Command to generate one reference
- [x] Create a Command to generate all references (based on config file) - and their corresponding directories
- [ ] add command tests

### CODE REVIEW

File: docs/src/Command/GenerateReferenceCommand.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    private string $root;
+    private \ReflectionClass $reflectionClass;
+
+    public function __construct(
+        private readonly PhpDocHelper $phpDocHelper,
+        private readonly ReflectionHelper $reflectionHelper,
+        string $name = null
+    ) {
+        parent::__construct($name);
+        $this->config = (require 'src/config.php')();
```
Comment (soyuka) [inline docs/src/Command/GenerateReferenceCommand.php:39]: At some point we should put this in a bundle using a symfony Configuration?

File: docs/src/Command/GenerateReferenceCommand.php [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        $this->root = Path::makeAbsolute($this->config['reference']['src'], getcwd());
+    }
+
+    public function openFile(mixed $fileName): void
+    {
+        $handle = fopen($fileName, 'r');
+        if (!$handle) {
+            fwrite(\STDERR, sprintf('Error opening %s. %s', $fileName, \PHP_EOL));
```
Comment (soyuka) [inline docs/src/Command/GenerateReferenceCommand.php:47]: use this to get stderr:

```
        $io = new SymfonyStyle($input, $output);
        $stderr = $io->getErrorStyle();
```

File: docs/src/Command/GenerateReferenceCommand.php [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                'leave empty for automatic generation, use `stdout` for printing'
+            );
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $style = new SymfonyStyle($input, $output);
+        $fileName = $input->getArgument('filename');
+
+        $this->openFile($fileName);
```
Comment (soyuka) [inline docs/src/Command/GenerateReferenceCommand.php:71]: really not sure you need a function for this

File: docs/src/Command/GenerateReferenceCommand.php [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        $file = Path::makeAbsolute($fileName, getcwd());
+        $relative = Path::makeRelative($file, $this->root);
+
+        $style->info(sprintf('Generating reference for %s', $relative));
+        $namespace = 'ApiPlatform\\'.str_replace(['/', '.php'], ['\\', ''], $relative);
+        $content = '';
+
+        $this->reflectionClass = new \ReflectionClass($namespace);
+        $outputFile = match ($outputArg = $input->getArgument('output')) {
+            null => $this->getOutPutFile($file),
```
Comment (soyuka) [inline docs/src/Command/GenerateReferenceCommand.php:82]: we should let the naming part to another script, I think its out of the responsability of this one. Let's leave this here for now we'll refactor this later.

File: docs/src/Command/GenerateReferenceCommand.php [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        $content = $this->writePageTitle($content);
+        $content = $this->writeClassName($content);
+        $content = $this->reflectionHelper->handleParent($this->reflectionClass, $content);
+        $content = $this->reflectionHelper->handleImplementations($this->reflectionClass, $content);
+        $content = $this->phpDocHelper->handleClassDoc($this->reflectionClass, $content);
+        $content = $this->reflectionHelper->handleClassConstants($this->reflectionClass, $content);
+        $content = $this->reflectionHelper->handleProperties($this->reflectionClass, $content);
+        $content = $this->reflectionHelper->handleMethods($this->reflectionClass, $content);
+
+        if (\STDOUT !== $outputFile) {
```
Comment (soyuka) [inline docs/src/Command/GenerateReferenceCommand.php:96]: I think we can use $output but let's keep this for later (see above comment)

File: docs/src/Services/Reference/PhpDocHelper.php [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        }
+
+        return $propertiesConstructorDocumentation;
+    }
+
+    public function handleClassDoc(\ReflectionClass $class, string $content): string
+    {
+        $rawDocNode = $class->getDocComment();
+
+        if ($rawDocNode) {
```
Comment (soyuka) [inline docs/src/Services/Reference/PhpDocHelper.php:56]: use a guard clause:

```
if (!$rawDocNode) {
    return $content;
}
```

File: docs/src/Command/GenerateReferenceCommand.php (hunk 1 above)
Comment (JacquesDurand) [reply docs/src/Command/GenerateReferenceCommand.php:39]: Yeah it is actually troublesome now that I'm trying to add tests for the commands. (Paths do not match etc ...)

What do you think would be the best way to handle this ?

File: docs/src/Services/Reference/ReflectionHelper.php [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+            $content .= '### Extends: '.\PHP_EOL;
+            $content .= '> '.$this->outputFormatter->addLink($parent).\PHP_EOL;
+        }
+
+        return $content;
+    }
+
+    public function handleImplementations(\ReflectionClass $reflectionClass, string $content): string
+    {
+        if (!empty($reflectionClass->getInterfaces())) {
```
Comment (soyuka) [inline docs/src/Services/Reference/ReflectionHelper.php:39]: use a guard clause:

```
if (!$interfaces = $reflectionClass->getInterfaces()) {
    return $content;
}

$content .= '### Implements '.\PHP_EOL;

foreach ($interfaces as $interface) {
    $content .= '> '.$this->outputFormatter->addLink($interface).\PHP_EOL.'> '.\PHP_EOL;
}
```

File: docs/src/Command/GenerateReferenceCommand.php (hunk 3 above)
Comment (JacquesDurand) [reply docs/src/Command/GenerateReferenceCommand.php:71]: Now that I think about it, I'm not sure we even need actually opening the file at all ?

File: docs/src/Services/Reference/ReflectionHelper.php [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+        return $content;
+    }
+
+    public function handleClassConstants(\ReflectionClass $reflectionClass, string $content): string
+    {
+        $constants = [];
+        foreach ($reflectionClass->getReflectionConstants(\ReflectionClassConstant::IS_PUBLIC) as $constant) {
+            $constants[] = $constant;
+        }
+        if (!empty($constants)) {
```
Comment (soyuka) [inline docs/src/Services/Reference/ReflectionHelper.php:56]: is `empty` really needed?

File: docs/src/Services/Reference/ReflectionHelper.php [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+                $content .= $text.\PHP_EOL;
+            }
+        }
+
+        return $content;
+    }
+
+    public function handleProperties(\ReflectionClass $reflectionClass, string $content): string
+    {
+        if (!empty($reflectionClass->getProperties())) {
```
Comment (soyuka) [inline docs/src/Services/Reference/ReflectionHelper.php:86]:
```
if (!$properties = $reflectionClass->getProperties()) {}

foreach ($properties as
```

File: docs/src/Services/Reference/ReflectionHelper.php [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+    public function handleMethods(\ReflectionClass $reflectionClass, string $content): string
+    {
+        $methods = [];
+        foreach ($reflectionClass->getMethods() as $method) {
+            if (!$this->hasToBeSkipped($method, $reflectionClass)) {
+                $methods[] = $method;
+            }
+        }
+
+        if (!empty($methods)) {
```
Comment (soyuka) [inline docs/src/Services/Reference/ReflectionHelper.php:191]: if (!$methods)

File: docs/src/Services/Reference/ReflectionHelper.php [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+    }
+
+    public function containsOnlyPrivateMethods(\ReflectionClass $reflectionClass): bool
+    {
+        // Do not skip empty interfaces
+        if (interface_exists($reflectionClass->getName()) || trait_exists($reflectionClass->getName())) {
+            return false;
+        }
+
+        if (!empty($reflectionClass->getProperties())) {
```
Comment (soyuka) [inline docs/src/Services/Reference/ReflectionHelper.php:353]:
```suggestion
        if (!$reflectionClass->getProperties()) {
```

not sure empty is needed

**Pull request merged**